src/audio/chain_dma.c

```
@@ -37,56 +48,478 @@ static struct comp_dev *chain_dma_create(const struct comp_driver *drv,
+	if (ret)
+		goto error;
+
+	cd->chain_task.state = SOF_TASK_STATE_INIT;
+	k_spin_unlock(&drivers->lock, key);
+
 	return IPC4_SUCCESS;
+
+error:
+	k_spin_unlock(&drivers->lock, key);
+	return IPC4_FAILURE;
 }
 
-int chain_dma_pause(struct comp_dev *dev)
+int chain_task_pause(struct comp_dev *dev)
 {
+	struct chain_dma_data *cd = comp_get_drvdata(dev);
+	struct comp_driver_list *drivers = comp_drivers_get();
+	k_spinlock_key_t key;
+
+	if (cd->chain_task.state == SOF_TASK_STATE_FREE)
+		return IPC4_SUCCESS;
+
+	key = k_spin_lock(&drivers->lock);
+	cd->first_data_received = false;
+
+	if (chain_stop(dev) < 0) {
+		k_spin_unlock(&drivers->lock, key);
+		return IPC4_FAILURE;
+	}
+
+	cd->chain_task.state = SOF_TASK_STATE_COMPLETED;
+	k_spin_unlock(&drivers->lock, key);
+	schedule_task_free(&cd->chain_task);
+	pm_policy_state_lock_put(PM_STATE_RUNTIME_IDLE, PM_ALL_SUBSTATES);
+
 	return IPC4_SUCCESS;
 }
 
-int chain_dma_trigger(struct comp_dev *dev, int cmd)
+int chain_task_trigger(struct comp_dev *dev, int cmd)
 {
 	switch (cmd) {
 	case COMP_TRIGGER_START:
-		return chain_dma_start(dev);
+		return chain_task_start(dev);
 	case COMP_TRIGGER_PAUSE:
-		return chain_dma_pause(dev);
+		return chain_task_pause(dev);
 	default:
 		return IPC4_ERROR_INVALID_PARAM;
 	}
 }
 
-void chain_dma_remove(struct comp_dev *dev)
+void chain_task_free(struct comp_dev *dev)
 {
 	struct chain_dma_data *cd = comp_get_drvdata(dev);
 
+	chain_release(dev);
 	rfree(cd);
 	rfree(dev);
 }
 
+static enum task_state chain_task_run(void *data)
+{
+	struct dma_status stat;
+	size_t link_avail_bytes, link_free_bytes, host_avail_bytes, host_free_bytes;
+	uint32_t link_read_pos, host_read_pos;
+	struct chain_dma_data *cd = (struct chain_dma_data *)data;
+
+	/* retrieve current status of link DMA */
+	int ret = dma_get_status(cd->chan_link->dma->z_dev, cd->chan_link->index,
+				 &stat);
+	if (ret < 0)
+		tr_err(&chain_dma_tr, "chain_task_run(): dma_get_status() failed, ret = %u",
+			 ret);
```

alignment to parenthesis

---

So, it is an error, yet we continue? So it is not an error, or?

---

It could return error when xrun is detected. Then the task should report an error and continue with flow.